packages/tldraw/src/lib/ui/components/Toolbar/DefaultImageToolbar.tsx

```diff
@@ -0,0 +1,135 @@
+	if (!imageShape || !showToolbar) return null
+
+	return (
+		<ContextualToolbarInner imageShape={imageShape as TLImageShape}>
+			{children}
+		</ContextualToolbarInner>
+	)
+})
+
+function ContextualToolbarInner({
+	children,
+	imageShape,
+}: {
+	children?: React.ReactNode
+	imageShape: TLImageShape
+}) {
+	const editor = useEditor()
+	const editorPath = useValue('editor path', () => editor.getPath(), [editor])
+	const isInCropTool = editorPath.startsWith('select.crop.')
+	const isCropping = editorPath === 'select.crop.cropping'
+	const [isEditingAltText, setIsEditingAltText] = useState(false)
+	const [isManipulating, setIsManipulating] = useState(false)
+	const [cachedManipulatingScreenBounds, setCachedManipulatingScreenBounds] = useState<
+		Box | undefined
+	>(undefined)
+	const [forceRerender, setForceReRender] = useState(0)
+	const handleEditAltTextStart = useCallback(() => {
+		setIsEditingAltText(true)
+		setForceReRender((n) => n + 1)
+	}, [])
+	const handleManipulatingStart = useCallback(() => {
+		setIsManipulating(true)
+		editor.setCurrentTool('select.crop.idle')
+		setForceReRender((n) => n + 1)
+	}, [editor])
+	const handleManipulatingEnd = useCallback(() => {
+		setIsManipulating(false)
+		editor.setCurrentTool('select.idle')
+		setForceReRender((n) => n + 1)
+	}, [editor])
+	const onEditAltTextComplete = useCallback(() => {
+		setIsEditingAltText(false)
+		setForceReRender((n) => n + 1)
+	}, [])
+
+	const getSelectionBounds = useCallback(() => {
+		const fullBounds = editor.getSelectionScreenBounds()
+		if (!fullBounds) return undefined
+		const bounds = new Box(fullBounds.x, fullBounds.y, fullBounds.width, 0)
+
+		// We cache the bounds because when manipulating it's annoying that the
+		// toolbar moves around, esp. when rotating.
+		if (isManipulating) {
+			if (!cachedManipulatingScreenBounds) {
+				setCachedManipulatingScreenBounds(bounds)
+			} else {
+				return cachedManipulatingScreenBounds
+			}
+		} else {
+			setCachedManipulatingScreenBounds(undefined)
+		}
+		return bounds
+	}, [editor, isManipulating, cachedManipulatingScreenBounds])
+
+	useQuickReactor(
+		'camera position',
+		function updateToolbarPositionAndDisplay() {
+			// capture / force this to update when the camera moves
+			editor.getCamera()
+			setCachedManipulatingScreenBounds(undefined)
+		},
+		[editor]
+	)
+
+	useEffect(() => {
+		if (isInCropTool && !isManipulating) {
+			handleManipulatingStart()
+		}
+		if (!isInCropTool && isManipulating) {
+			handleManipulatingEnd()
+		}
+	}, [isInCropTool, editor, isManipulating, handleManipulatingStart, handleManipulatingEnd])
+
+	if (isCropping) return null
+
+	return (
+		<TldrawUiContextualToolbar
+			// TODO: this is a little hack to force the toolbar to re-render when
+			// we go into crop mode.
+			key={forceRerender}
+			className="tlui-image__toolbar"
+			getSelectionBounds={getSelectionBounds}
+		>
+			{children ? (
+				children
+			) : isEditingAltText ? (
+				<AltTextEditor
```

Comment:
it's a bit surprising to me that this isn't part of the content. why is that?

Reply:
Looking at this a bit more closely i'm less clear that the Thing/ThingContent pattern makes sense here given how tightly coupled these are. Although possibly if we did the resize observer thing we'd be able to get rid of a bunch of that coupling as this component would no longer need to know when its child might change size?

Reply:
(see above on` key={forceRerender}` comment)

Reply:
Alex addressed in this PR 🎉 
https://github.com/tldraw/tldraw/pull/5954